include/ldsCtrlEst_h/lds_mpcctrl.h

```diff
@@ -0,0 +1,533 @@
+    Vector Qxr = Qxr_full.as_col();  // Qxr for every simulation time step
+
+    Vector qu =
+        join_vert((-2 * S_ * u0), Vector((M_ - 1) * m_, arma::fill::zeros));
+    Vector qx = Qxr.rows(0, N_ * n_ - 1);
+    q_arma = join_vert(qx, qu);
+
+  }
+
+  // Problem values for OSQP
+  OSQPFloat* q = from_vec(q_arma);
+  OSQPFloat* lb = from_vec(lb_);
+  OSQPFloat* ub = from_vec(ub_);
+
+  // Solver, settings, matrices
+  OSQPSolver* solver;
+  OSQPSettings* settings;
+  OSQPCscMatrix* A = from_matrix(Acon_);
+  OSQPCscMatrix* P = from_matrix(P_);
+
+  // Set settings
+  settings = (OSQPSettings*)malloc(sizeof(OSQPSettings));
+  if (settings) {
+    osqp_set_default_settings(settings);
+    settings->verbose = false;
+    settings->warm_starting = false;
+  }
+
+  // Set up solver
+  OSQPInt exit_flag = osqp_setup(&solver, P, q, A, lb, ub, A->m, A->n, settings); // n and m are set by A cols and rows
+
+  // Solve problem
+  if (!exit_flag) exit_flag = osqp_solve(solver);
+
+  if (solver->info->status_val != OSQP_SOLVED) {
+    std::string error_message = "OSQP solver failed to solve.";
+    throw std::runtime_error(error_message);
+  }
+
+  OSQPSolution* sol = std::move(solver->solution);
+
+  // Clean up
+  osqp_cleanup(solver);
+  if (q) free(q);
+  if (lb) free(lb);
+  if (ub) free(ub);
+  if (A->i) free(A->i);
+  if (A->p) free(A->p);
+  if (A->x) free(A->x);
+  if (A) free(A);
+  if (P->i) free(P->i);
+  if (P->p) free(P->p);
+  if (P->x) free(P->x);
+  if (P) free(P);
+  if (settings) free(settings);
+
+  return sol;
+}
+
+template <typename System>
+OSQPCscMatrix* MpcController<System>::from_matrix(const Matrix& A) {
+  OSQPCscMatrix* mat = (OSQPCscMatrix*)malloc(sizeof(OSQPCscMatrix));
+
+  mat->m = A.n_rows;
+  mat->n = A.n_cols;
+  mat->nzmax = A.n_rows * A.n_cols;  // Can be made smaller if needed
+  mat->nz = -1;  // -1 means the matrix is in CSC format (required for API)
+
+  mat->p = (OSQPInt*)malloc((mat->n + 1) * sizeof(OSQPInt));
+  mat->i = (OSQPInt*)malloc(mat->nzmax * sizeof(OSQPInt));
+  mat->x = (OSQPFloat*)malloc(mat->nzmax * sizeof(OSQPFloat));
+  
+  if (!mat->p || !mat->i || !mat->x) {
+    std::string error_message = "Failed to allocate memory for OSQP matrix.";
+    throw std::runtime_error(error_message);
+  }
+
+  OSQPInt n = 0;
+  mat->p[0] = 0;
+  for (OSQPInt j = 0; j < mat->n; j++) {
+    for (OSQPInt i = 0; i < mat->m; i++) {
+      if (A(i, j) != 0) {
+        mat->i[n] = i;
+        mat->x[n] = A(i, j);
+        n++;
+      }
+    }
+    mat->p[j + 1] = n;
+  }
+
+  return mat;
+}
+
+template <typename System>
+OSQPCscMatrix* MpcController<System>::from_sparse(
+    const arma::SpMat<data_t>& A) {
+  OSQPCscMatrix* mat = (OSQPCscMatrix*)malloc(sizeof(OSQPCscMatrix));
+
+  mat->m = A.n_rows;
+  mat->n = A.n_cols;
+  mat->nzmax = A.n_nonzero;
+  mat->nz = -1;  // -1 means the matrix is in CSC format (required for API)
+
+  mat->p = (OSQPInt*)malloc((mat->n + 1) * sizeof(OSQPInt));
```

Comment:
again, hard to see if there's a potential memory bug here when I don't know what p, i, x are

Reply:
I will fix that. Those are variables associated with the `OSQPCscMatrix` type. 

> **OSQPInt \*p**[¶](https://osqp.org/docs/interfaces/C.html#_CPPv4N13OSQPCscMatrix1pE)
> column pointers (size n+1); col indices (size nzmax) starting from 0 for triplet format
> 
> **OSQPInt \*i**[¶](https://osqp.org/docs/interfaces/C.html#_CPPv4N13OSQPCscMatrix1iE)
> row indices, size nzmax starting from 0
> 
> **OSQPFloat \*x**[¶](https://osqp.org/docs/interfaces/C.html#_CPPv4N13OSQPCscMatrix1xE)
> numerical values, size nzmax